core/discovery/storage.go

```diff
@@ -36,16 +53,91 @@ type ProposalStorage struct {
 	mutex     sync.RWMutex
 }
 
-// Proposals returns table of proposals in storage
+// Proposals returns list of proposals in storage
 func (storage *ProposalStorage) Proposals() []market.ServiceProposal {
 	storage.mutex.Lock()
 	defer storage.mutex.Unlock()
+
 	return storage.proposals
 }
 
-// Set puts given proposal to storage
-func (storage *ProposalStorage) Set(proposals ...market.ServiceProposal) {
+// MatchProposals fetches currently active service proposals from discovery by match function
+func (storage *ProposalStorage) MatchProposals(match ProposalReducer) ([]market.ServiceProposal, error) {
```

Comment:
Prefer to use short receiver name like `func (s *ProposalStorage)`. It is easer to read, otherwise storage looks like some local variable or package name.